src/Microsoft.AspNetCore.Server.IISIntegration/Server/IISHttpContext.cs

```diff
@@ -418,279 +429,264 @@ public unsafe void SendResponseHeaders(bool appCompleted)
             }
         }
 
+        // Needs to be locked when accessing this method.
+        private void ModifyReadCancellationState()
+        {
+            switch (_currentOperationType)
+            {
+                case CurrentOperationType.Read:
+                    NativeMethods.http_cancel_io(_pInProcessHandler);
+                    break;
+                case CurrentOperationType.Write:
+                case CurrentOperationType.Flush:
+                case CurrentOperationType.None:
+                default:
+                    _allowedToRead = false;
+                    break;
+            }
+        }
+
         public void Abort()
         {
             // TODO
         }
 
-        public void StartReadingRequestBody()
+        public void StartProcessingRequestAndResponseBody()
         {
-            if (_readingTask == null)
+            if (_readWriteTask == null)
             {
-                _readingTask = ProcessRequestBody();
+                lock (_createReadWriteBodySync)
+                {
+                    if (_readWriteTask == null)
+                    {
+                        _readWriteTask = ProcessRequestAndResponseBody();
+                    }
+                }
             }
         }
 
-        private async Task ProcessRequestBody()
+        // IIS can only handle on async operation per request at a time.
+        // To guarantee this, we have a single loop to read and write the request/response
+        private async Task ProcessRequestAndResponseBody()
```

Comment:
This method is very long and pretty hard to follow (100 lines with 13 nested ifs). Worth refactoring/splitting into smaller methods? Or at least commenting a bit more.